pkg/compact/compact.go

```diff
@@ -320,37 +363,161 @@ func (g *DefaultGrouper) Groups(blocks map[ulid.ULID]*metadata.Meta) (res []*Gro
 	return res, nil
 }
 
+// concurrentGroups returns the compaction groups for all blocks currently known to the syncer. This differs from
+// the default grouping in that a single group will be split into a group per available plan.
+func (g *DefaultGrouper) concurrentGroups(blocks map[ulid.ULID]*metadata.Meta) (res []*Group, err error) {
+	groups := map[string]*Group{}
+	for _, m := range blocks {
+		if _, ok := g.locked[m.ULID]; ok {
+			continue
+		}
+		groupKey := DefaultGroupKey(m.Thanos)
+		group, ok := groups[groupKey]
+		if !ok {
+			lbls := labels.FromMap(m.Thanos.Labels)
+			group, err = NewGroup(
+				log.With(g.logger, "group", fmt.Sprintf("%d@%v", m.Thanos.Downsample.Resolution, lbls.String()), "groupKey", groupKey),
+				g.bkt,
+				groupKey,
+				"",
+				lbls,
+				m.Thanos.Downsample.Resolution,
+				g.acceptMalformedIndex,
+				g.enableVerticalCompaction,
+				g.enableConcurrentGroupCompaction,
+				g.compactions.WithLabelValues(groupKey),
+				g.compactionRunsStarted.WithLabelValues(groupKey),
+				g.compactionRunsCompleted.WithLabelValues(groupKey),
+				g.compactionFailures.WithLabelValues(groupKey),
+				g.verticalCompactions.WithLabelValues(groupKey),
+				g.garbageCollectedBlocks,
+				g.blocksMarkedForDeletion,
+				g.hashFunc,
+			)
+			if err != nil {
+				return nil, errors.Wrap(err, "create compaction group")
+			}
+			groups[groupKey] = group
+			res = append(res, group)
+		}
+		if err := group.Add(m); err != nil {
+			return nil, errors.Wrap(err, "add compaction group")
+		}
+	}
+
+	newGroups := []*Group{}
+	// Split each group by available plans.
+	for _, group := range res {
+		newGroup, err := g.splitGroupByPlans(group)
+		if err != nil {
+			return nil, err
+		}
+		newGroups = append(newGroups, newGroup...)
+	}
+	res = newGroups
+
+	sort.Slice(res, func(i, j int) bool {
+		return res[i].Key() < res[j].Key()
+	})
+	return res, nil
+}
+
+// Recursively plan a groups metas until there are no available plans and return a new group for each plan.
+// This allows a single group key to utilize group concurrency when there are multiple plans available to compact.
+func (g *DefaultGrouper) splitGroupByPlans(group *Group) ([]*Group, error) {
+	groups := []*Group{}
+	var seededRand = rand.New(rand.NewSource(time.Now().UnixNano()))
```

Comment:
This catches my attention :D  Side-note here; I'm not an expert on this part, so I could be totally wrong. ^^

Should a seedRand only be initialised once? Since this func is called multiple times from other loops. Then we can use https://golang.org/pkg/math/rand/#Intn to get a new pseudo-int. Which is I believe faster/easier. Especially if we do a loop that might (??) be done in nano second, you'll end up with the same seed.

To continue about easier; what about ulid for the hash? I'm not worried about collisions, but it's pretty easy and consistent with the usage of ulid for blocks. 

Not sure if relevant, but the oklog/ulid package has this statement too; 

> Please note that rand.Rand from the math package is not safe for concurrent use. Instantiate one per long living go-routine or use a sync.Pool if you want to avoid the potential contention of a locked rand.Source as its been frequently observed in the package level functions.

Reply:
good call! I removed the hash completely and instead just used the first ULID in the compaction group to append to the sub directory name which also cleaned up needing hash property on group.